paf-mvp-operator-express/src/operator-node.ts

```diff
@@ -76,443 +73,536 @@ export interface OperatorNodeConfig extends Config {
-          type: OperatorErrorType.UNKNOWN_ERROR,
-          details: '',
-        };
-        res.status(400);
-        res.json(error);
-      }
-    });
-
-    app.expressApp.get(jsonOperatorEndpoints.newId, cors(corsOptionsAcceptAll), async (req, res) => {
-      logger.Info(jsonOperatorEndpoints.newId);
-      const request = getPafDataFromQueryString<GetNewIdRequest>(req);
-      const context = { origin: req.header('origin') };
-
-      const sender = request.sender;
-
-      if (!allowedHosts[sender]?.includes(Permission.READ)) {
-        throw `Domain not allowed to read data: ${sender}`;
-      }
+      res.status(400);
+      res.json(error);
+      next(error);
+    }
+  }
 
-      try {
-        if (
-          !(await getNewIdRequestVerifier.verifySignatureAndContent(
-            { request, context },
-            sender, // sender will always be ok
-            operatorHost // but operator needs to be verified
-          ))
-        ) {
-          // TODO [errors] finer error feedback
-          throw 'New Id request verification failed';
-        }
-
-        const response = getNewIdResponseBuilder.buildResponse(request.receiver, idBuilder.generateNewId());
-        res.json(response);
-      } catch (e) {
-        logger.Error(jsonOperatorEndpoints.newId, e);
-        // FIXME finer error return
-        const error: OperatorError = {
-          type: OperatorErrorType.UNKNOWN_ERROR,
-          details: '',
-        };
-        res.status(400);
-        res.json(error);
-      }
-    });
+  async redirectRead(req: Request, res: Response, next: NextFunction) {
+    const request = getPafDataFromQueryString<RedirectGetIdsPrefsRequest>(req);
 
-    // *****************************************************************************************************************
-    // ******************************************************************************************************* REDIRECTS
-    // *****************************************************************************************************************
+    if (!request?.returnUrl) {
+      // FIXME more robust error handling: websites should not be broken in this case, do a redirect with empty data
+      const error: OperatorError = {
+        type: OperatorErrorType.INVALID_RETURN_URL,
+        details: '',
+      };
+      res.status(400);
+      res.json(error);
+      next(error);
+      return;
+    }
+
+    try {
+      const response = await this.getReadResponse(request, req);
+
+      const redirectResponse = this.getIdsPrefsResponseBuilder.toRedirectResponse(response, 200);
+      const redirectUrl = this.getIdsPrefsResponseBuilder.getRedirectUrl(new URL(request?.returnUrl), redirectResponse);
+
+      httpRedirect(res, redirectUrl.toString());
+      next();
+    } catch (e) {
+      this.logger.Error(redirectEndpoints.read, e);
+      // FIXME more robust error handling: websites should not be broken in this case, do a redirect with empty data
```

Comment:
This code is repeated in different functions, What do you think about factorizing it ?

Reply:
Yes, in the PR focused on error handling.

Reply:
Ok.